wrapanapi/lenovo.py

```diff
@@ -1,42 +1,309 @@
+        """An instance of the service"""
+        try:
+            response = requests.put(self.url + path, data=json.dumps(request), auth=self.auth,
+                    verify=False)
+            return response
+        except Timeout:
+            return None
 
     @property
     def version(self):
         """The product version"""
-        raise NotImplementedError
+        response = self._service_instance("aicc")
+        return response['appliance']['version']
+
+    def list_servers(self):
+        response = self._service_instance("cabinet?status=includestandalone")
+        cabinets = response['cabinetList'][0]
+        nodes_list = cabinets['nodeList']
+
+        inventory = [node['itemInventory'] for node in nodes_list] if len(nodes_list) > 0 else []
+
+        nodes_from_chassis = []
+        if len(cabinets['chassisList']) > 0:
+            chassis_list = cabinets['chassisList'][0]
+            nodes_from_chassis = [node for node in chassis_list['itemInventory']['nodes']
+                            if node['type'] != 'SCU']
+
+        inventory.extend(nodes_from_chassis)
+
+        self._servers_list = inventory
+        return inventory
+
+    def change_node_power_status(self, server, request):
+        url = "nodes/" + str(server['uuid'])
+        payload = {'powerState': request}
+        response = self._service_put(url, payload)
+
+        return response
+
+    def change_led_status(self, server, name, state):
+        url = "nodes/" + str(server['uuid'])
+        payload = {'leds': [{'name': name, 'state': state}]}
+        response = self._service_put(url, payload)
+
+        return response
+
+    def get_server(self, server_name):
+        if not self._servers_list:
+            self.list_servers()
+
+        try:
+            for node in self._servers_list:
+                if node['name'] == server_name:
+                    return node
+        except AttributeError:
+            return None
+
+    def get_led(self, server_name):
+        try:
+            server = self.get_server(server_name)
+            leds = server['leds']
+            for led in leds:
+                if led['name'] == 'Identify' or led['name'] == 'Identification':
+                    return led
+        except AttributeError:
+            return None
+
+    def get_server_hostname(self, server_name):
+        server = self.get_server(server_name)
+
+        return str(server['hostname'])
+
+    def get_server_ipv4_address(self, server_name, timeout=600):
+        server = self.get_server(server_name)
+
+        return server['ipv4Addresses']
+
+    def get_server_ipv6_address(self, server_name, timeout=600):
+        server = self.get_server(server_name)
+
+        return server['ipv6Addresses']
+
+    def get_server_mac_address(self, server_name, timeout=600):
+        server = self.get_server(server_name)
+
+        return server['macAddress']
+
+    def server_status(self, server_name):
+        server = self.get_server(server_name)
```

Comment:
@psav i beleive as a **followup** we should take a look at providing a extension point/mixing that handles most of this more nicely

as far as i can tell the number of similaryl implemented status/health methods as well as the is_...methods now justify providing some utility to deduplicate